include/envoy/router/router.h

```diff
@@ -173,9 +173,10 @@ class RetryPolicy {
   virtual std::chrono::milliseconds perTryTimeout() const PURE;
 
   /**
-   * @return uint32_t the number of retries to allow against the route.
+   * @return absl::optional<uint32_t> the number of retries to allow against the route.
+   *         Returns absl::nullopt if no retry count was specified in the route config.
    */
-  virtual uint32_t numRetries() const PURE;
+  virtual absl::optional<uint32_t> numRetries() const PURE;
```

Comment:
I'm tired and probably missing something obvious, but why do we need the optional here? Why can't the retries remaining just be set directly to 0 if this is what the route says, and then potentially overridden by headers?

Reply:
Its to maintain the current behavior in which setting the retry-on header when hitting a route without a retry block specified (the absl::nullopt case) we should default to 1 retry.

That said I think I have an easier implementation: just change the numRetries returned by the default constructed RetryPolicy to be one. Will push that up